services/base.js

```diff
@@ -219,19 +236,35 @@ class BaseService {
       },
     }));
   }
-};
+}
 
 class BaseJsonService extends BaseService {
-  async _requestJson(url, options = {}, notFoundMessage) {
+  static _validate(json, schema) {
+    const { error, value } = Joi.validate(json, schema, {
+      allowUnknown: true,
+      stripUnknown: true,
+    });
+    if (error) {
+      throw new InvalidResponse({
+        prettyMessage: 'invalid json response',
+        underlyingError: error,
+      });
+    } else {
+      return value;
+    }
+  }
+
+  async _requestJson({ schema, url, options = {}, notFoundMessage }) {
```

Comment:
Do you think it is worth explicitly checking `schema` has been passed here? If we pass `undefined` to `Joi.validate()`, it will throw `AssertionError [ERR_ASSERTION]: Invalid schema content:` but I wonder if it is worth throwing a 'friendlier' error in that case? I can see that being a common pitfall for new contributors. Would another way to 'enforce' this be for `schema` to be a class property instead of a function param. e.g:

```js
static schema() {
  throw new Error('must define schema');
}
```

Also, it seems like there will be a certain amount of cases where we do actually want to pass `Joi.any()` (it looks like that is what you are doing for `NpmLicense` and `NpmTypeDefinitions` in this PR). We definitely don't want that to be the _default_ case but is it worth providing some kind of 'shorthand' for that case?

Reply:
So, yea, I had the schema in as a class property before [(`static get responseSchema`)](https://github.com/badges/shields/blob/06a8dbf133c52c090ea12774aca50887e3b059be/services/node/node.js#L54-L56) but ended up making it a parameter. The schema compilation step is slow, so we don't want to compile a new schema every time we have a new request. Since we don't have class properties yet, that means we need to memoize it somehow, or else define it outside the class. Passing it in the call to `_requestJson` seems like less typing, is just as clear, and is possible to use conditionally.

`Joi.any()` seems like a good shorthand to me. Were you thinking something like `shouldValidate: false`? I'd be okay with that. However I don't want contributors to think that validating is optional. 😉 Maybe not having a shortcut is okay.

I'm good with adding a check with a better error message. Probably we should check that it's a Joi schema. I think Joi supports compiling from object literals, though constructing them all literally is self-documenting.

Reply:
> The schema compilation step is slow, so we don't want to compile a new schema every time we have a new request.

We've got enough performance problems at the moment. Lets not make that any worse ;)

> Maybe not having a shortcut is okay.

OK. Lets ignore that for now. Once we've refactored some more stuff we'll have an idea of how common this case is going to be.

> I'm good with adding a check with a better error message. Probably we should check that it's a Joi schema.

Yeah, lets ensure that `schema` exists and is an instance of `Joi` and throw an error which provides a hint if not.